rust/opendp/src/poly.rs

```diff
@@ -0,0 +1,117 @@
+use std::any::Any;
+
+use crate::core::{Domain, Function, Measure, Measurement, Metric, Transformation};
+use crate::error::*;
+
+/// A polymorphic Domain. This admits any value of any type (represented as a Box<dyn Any>).
+#[derive(PartialEq, Clone)]
+pub struct PolyDomain {}
+
+impl PolyDomain {
+    pub fn new() -> Self { PolyDomain {} }
+}
+
+impl Domain for PolyDomain {
+    type Carrier = Box<dyn Any>;
+    fn member(&self, _val: &Self::Carrier) -> bool { true }
+}
+
+impl<DI, DO> Function<DI, DO>
+    where DI: 'static + Domain,
+          DI::Carrier: 'static,
+          DO: 'static + Domain,
+          DO::Carrier: 'static {
+    /// Converts this Function into one with polymorphic output.
+    pub fn into_poly(self) -> Function<DI, PolyDomain> {
+        let function = move |arg: &DI::Carrier| -> Fallible<Box<dyn Any>> {
+            let res = (self.function)(arg);
+            res.map(|o| Box::new(*o) as Box<dyn Any>)
+        };
+        Function::new_fallible(function)
+    }
+}
+
+impl<DI: Domain> Function<DI, PolyDomain> {
```

Comment:
This is cool! `eval_poly` with the type arg to downcast to, exists on `Box<Any>` outputs.

Reply:
Yeah I think this should work as our general purpose composition type.